### ISSUE TITLE
NotImplementedError in RelativePositioningSampler._sample_pair

### ISSUE THREAD

**Issue by tgy** [braindecode/braindecode#452]
hi! i'm trying to use the `RelativePositioningSampler` on clinical EEG data

however i'm running into a `NotImplementedError` which I cannot explain from the code of `braindecode/samplers/ssl.py`

```python
def _sample_pair(self):
    """Sample a pair of two windows.
    """
    # Sample first window
    win_ind1, rec_ind1 = self.sample_window()
    ts1 = self.metadata.iloc[win_ind1]['i_start_in_trial']
    ts = self.info.iloc[rec_ind1]['i_start_in_trial']

    # Decide whether the pair will be positive or negative
    pair_type = self.rng.binomial(1, 0.5)
    win_ind2 = None
    if pair_type == 0:  # Negative example
        if self.same_rec_neg:
            mask = (
                ((ts <= ts1 - self.tau_neg) & (ts >= ts1 - self.tau_max)) |
                ((ts >= ts1 + self.tau_neg) & (ts <= ts1 + self.tau_max))
            )
        else:
            rec_ind2 = rec_ind1
            while rec_ind2 == rec_ind1:
                win_ind2, rec_ind2 = self.sample_window()
    elif pair_type == 1:  # Positive example
        mask = (ts >= ts1 - self.tau_pos) & (ts <= ts1 + self.tau_pos)

    if win_ind2 is None:
        mask[ts == ts1] = False  # same window cannot be sampled twice
        if sum(mask) == 0:
            raise NotImplementedError
        win_ind2 = self.rng.choice(self.info.iloc[rec_ind1]['index'][mask])

    return win_ind1, win_ind2, float(pair_type)
```

in particular this part of the code:

```python
    if win_ind2 is None:
        mask[ts == ts1] = False  # same window cannot be sampled twice
        if sum(mask) == 0:
            raise NotImplementedError
        win_ind2 = self.rng.choice(self.info.iloc[rec_ind1]['index'][mask])
```

the code generating this error:

```python
DATA_PATH = pathlib.Path("/data/mydataset")
random_state = 87
n_jobs = 12
window_size_s = 30
sfreq = 63
high_cut_hz = 31
factor = 1e6
window_size_samples = window_size_s * sfreq


class MyDataset(braindecode.datasets.base.BaseConcatDataset):    
    def __init__(self, preload=False): 
        paths = DATA_PATH.glob("*.raw.fif.gz")
        paths = list(paths)
        all_base_ds = list()
        for p in paths:
            raw, desc = self._load_raw(p, preload=preload)
            base_ds = braindecode.datasets.base.BaseDataset(raw, desc)
            all_base_ds.append(base_ds)
        super().__init__(all_base_ds)

    @staticmethod
    def _load_raw(raw_path: pathlib.Path, preload: bool):
        my_session_id = raw_path.stem
        raw = mne.io.read_raw_fif(raw_path, preload=preload)
        assert np.isclose(raw.info["sfreq"], sfreq)
        desc = pd.Series({"session": my_session_id}, name="")
        return raw, desc

dataset = MyDataset()

preprocessors = [
    braindecode.preprocessing.Preprocessor(lambda data: np.multiply(data, factor)),
    braindecode.preprocessing.Preprocessor(
        "filter", l_freq=None, h_freq=high_cut_hz, n_jobs=n_jobs
    ),
]
braindecode.preprocessing.preprocess(dataset, preprocessors)

windows_dataset = braindecode.preprocessing.create_fixed_length_windows(
    dataset,
    start_offset_samples=0,
    window_size_samples=window_size_samples,
    window_stride_samples=window_size_samples,
    drop_last_window=True,
)


braindecode.preprocessing.preprocess(
    windows_dataset,
    [
        braindecode.preprocessing.Preprocessor(
            sklearn.preprocessing.scale, channel_wise=True
        )
    ],
)

bed_stays = np.unique(windows_dataset.description["session"])
bs_train, bs_test = sklearn.model_selection.train_test_split(
    bed_stays, test_size=0.4, random_state=random_state
)
bs_valid, bs_test = sklearn.model_selection.train_test_split(
    bs_test, test_size=0.5, random_state=random_state
)

class RelativePositioningDataset(braindecode.datasets.BaseConcatDataset):   
    def __init__(self, list_of_ds):  
        super().__init__(list_of_ds)
        self.return_pair = True

    def __getitem__(self, index):   
        if self.return_pair:
            ind1, ind2, y = index
            return (super().__getitem__(ind1)[0], super().__getitem__(ind2)[0]), y
        else:
            return super().__getitem__(index)

    @property
    def return_pair(self): 
        return self._return_pair

    @return_pair.setter
    def return_pair(self, value):
        self._return_pair = value


# %%

split_ids = {"train": bs_train, "valid": bs_valid, "test": bs_test}
splitted = dict()
for name, values in split_ids.items():
    splitted[name] = RelativePositioningDataset(
        [ds for ds in windows_dataset.datasets if ds.description["session"] in values]
    )


tau_pos, tau_neg = int(sfreq * 60), int(sfreq * 15 * 60)
n_examples_train = 250 * len(splitted["train"].datasets)
n_examples_valid = 250 * len(splitted["valid"].datasets)
n_examples_test = 250 * len(splitted["test"].datasets)

train_sampler = braindecode.samplers.RelativePositioningSampler(
    splitted["train"].get_metadata(),
    tau_pos=tau_pos,
    tau_neg=tau_neg,
    n_examples=n_examples_train,
    same_rec_neg=True,
    random_state=random_state,
)
valid_sampler = braindecode.samplers.RelativePositioningSampler(
    splitted["valid"].get_metadata(),
    tau_pos=tau_pos,
    tau_neg=tau_neg,
    n_examples=n_examples_valid,
    same_rec_neg=True,
    random_state=random_state,
).presample()
test_sampler = braindecode.samplers.RelativePositioningSampler(
    splitted["test"].get_metadata(),
    tau_pos=tau_pos,
    tau_neg=tau_neg,
    n_examples=n_examples_test,
    same_rec_neg=True,
    random_state=random_state,
).presample()
```

any idea of what could be happening? cc @agramfort @hubertjb @DavidSabbagh

**Comment by hubertjb** [comment]
Hello @tgy! Can you check the length of `ts` (and therefore of `mask`)? One possibility is that (at least) one of the recordings is too short for the `tau_neg` that you use, such that you might end up with no available negative windows. With your current `tau_neg` you'd need recordings of at least 63 * 15 * 60 * 2 = 113,400 samples to be sure not to fall in that case.

**Comment by tgy** [comment]
thanks a lot @hubertjb that was exactly it! :)
